### PR TITLE
VPN-2529: Fetch recommended servers

### PR DESCRIPTION
**Pull request by oskirby**
## Description
This implements a method to return a list of recommended servers, `ServerCountryModel::recommendedServers()` which returns a list of `ServerCity` objects. Along the way we also tackled some cleanup:
- `cityConnectionScore()` has been converted to a property of the `ServerCity` object, allowing QML to update itself as scores change.
- Move the great-circle distance algorithm into the `Location` model.
- Trigger server latency updates on changes to the underlying features.
- Sort the order of server latency fetches by geographic distance, to try and get the best-performing results first.
- Make `ServerCountryModel::pickBest()` a wrapper for `ServerCountryModel::recommendedServers(1)`

## Reference
Github #3988 ([VPN-2529](https://mozilla-hub.atlassian.net/browse/VPN-2529))
Github #4966 ([VPN-3297](https://mozilla-hub.atlassian.net/browse/VPN-3297))

## Checklist
    
- [ ] My code follows the style guidelines for this project
- [ ] I have not added any packages that contain high risk or unknown licenses (GPL,  LGPL, MPL, etc. consult with DevOps if in question)
- [ ] I have performed a self review of my own code
- [ ] I have commented my code PARTICULARLY in hard to understand areas
- [ ] I have added thorough tests where needed

[VPN-2529]: https://mozilla-hub.atlassian.net/browse/VPN-2529?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[VPN-3297]: https://mozilla-hub.atlassian.net/browse/VPN-3297?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: nebula/mocks/Mozilla/VPN/MZFilterProxyModel.qml [hunk 1]
```diff
@@ -0,0 +1,12 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import QtQuick 2.0
+
+Item {
+  property var filterCallback: () => {}
```
Comment (bakulf) [inline nebula/mocks/Mozilla/VPN/MZFilterProxyModel.qml:8]: This should not required. Why do we need to lock this object? Is it for qmllint?

File: src/apps/vpn/models/servercountrymodel.cpp [hunk 2]
```diff
@@ -364,10 +302,54 @@ void ServerCountryModel::retranslate() {
   endResetModel();
 }
 
+unsigned int ServerCountryModel::avgLatency() const {
+  if (m_numLatencySamples == 0) {
+    return 0;
+  } else {
```
Comment (bakulf) [inline src/apps/vpn/models/servercountrymodel.cpp:308]: No else here

File: src/apps/vpn/models/servercountrymodel.cpp [hunk 3]
```diff
@@ -364,10 +302,54 @@ void ServerCountryModel::retranslate() {
...
+  if (server.latency() != 0) {
+    m_sumLatencyMsec -= server.latency();
+  } else {
+    m_numLatencySamples++;
+  }
+  m_sumLatencyMsec += msec;
+  server.setLatency(msec);
+
+  // Ugly iteration. Find the city for this server.
+  // TODO: Maybe it would be cleaner to move the latency data into ServerCity.
```
Comment (bakulf) [inline src/apps/vpn/models/servercountrymodel.cpp:329]: Or move this servers into the cities.

**Review by bakulf — APPROVED** [review]

File: nebula/mocks/Mozilla/VPN/MZFilterProxyModel.qml (hunk 1 above)
Comment (oskirby) [reply nebula/mocks/Mozilla/VPN/MZFilterProxyModel.qml:8]: Yes, this was an attempt to clean up some qmllint warnings. It was complaining that it couldn't determine the types of properties in an `MZFilterProxyModel`, so I simply attempted to add a mocked version. Unfortunately, there are a bunch of other QML linter warnings that are going to much harder to resolve. I'm debating if I should just back out of the linter work and deal with it in another PR.

File: src/apps/vpn/models/servercountrymodel.cpp (hunk 3 above)
Comment (oskirby) [reply src/apps/vpn/models/servercountrymodel.cpp:329]: The problem is that we also have a need to frequently look up servers by their public key, so moving the `Servers` back into the `ServerCity` model  would just cause some ugly iteration elsewhere. What we almost need is something like a database where both servers and cities can be searched via multiple criteria.

File: nebula/mocks/Mozilla/VPN/MZFilterProxyModel.qml (hunk 1 above)
Comment (oskirby) [reply nebula/mocks/Mozilla/VPN/MZFilterProxyModel.qml]: @strseb  is at work on a much better solution with QML modules. So I am going to revert this.

**Review by lesleyjanenorton — APPROVED** [review]

**Pull request merged**